File: image-crud/src/main/java/org/judexmars/imagecrud/config/kafka/KafkaConfiguration.java
package org.judexmars.imagecrud.config.kafka;

import static org.springframework.kafka.core.KafkaAdmin.NewTopics;

import java.util.Map;
import java.util.function.Consumer;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.RoundRobinAssignor;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.RoundRobinPartitioner;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.kafka.common.serialization.StringSerializer;
import org.judexmars.imagecrud.dto.kafka.ImageStatusMessage;
import org.springframework.boot.autoconfigure.kafka.KafkaProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory;
import org.springframework.kafka.config.TopicBuilder;
import org.springframework.kafka.core.ConsumerFactory;
import org.springframework.kafka.core.DefaultKafkaConsumerFactory;
import org.springframework.kafka.core.DefaultKafkaProducerFactory;
import org.springframework.kafka.core.KafkaAdmin;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.core.ProducerFactory;
import org.springframework.kafka.listener.ContainerProperties;
import org.springframework.kafka.listener.DefaultErrorHandler;
import org.springframework.kafka.support.serializer.JsonDeserializer;
import org.springframework.kafka.support.serializer.JsonSerializer;
import org.springframework.util.backoff.FixedBackOff;

/**
 * Kafka configuration.
 */
@Configuration
@Slf4j
@RequiredArgsConstructor
public class KafkaConfiguration {

  private final KafkaProperties properties;

  /**
   * Register all required topics.
   *
   * @return kafka topics
   */
  @Bean
  public NewTopics topics() {
    return new KafkaAdmin.NewTopics(
        TopicBuilder.name("images.wip")
            .replicas(3)
            .partitions(2)
            .build(),
        TopicBuilder.name("images.done")
            .replicas(3)
            .partitions(2)
            .build());
  }

  @Bean("IMAGES_WIP_TEMPLATE")
  public KafkaTemplate<String, ImageStatusMessage> kafkaTemplate() {
    return new KafkaTemplate<>(producerFactory(props ->
        props.put(ProducerConfig.ACKS_CONFIG, "all")));
  }

  private <K, V> ProducerFactory<K, V> producerFactory(
      Consumer<Map<String, Object>> enchanter) {
    var props = properties.buildProducerProperties(null);

    props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
    props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, JsonSerializer.class);

    props.put(ProducerConfig.PARTITIONER_CLASS_CONFIG, RoundRobinPartitioner.class);

    props.put(ProducerConfig.LINGER_MS_CONFIG, 0);

    props.put(ProducerConfig.RETRIES_CONFIG, 3);

    enchanter.accept(props);

    return new DefaultKafkaProducerFactory<>(props);
  }

  /**
   * Factory for kafka consumer.
   *
   * @return {@link ConcurrentKafkaListenerContainerFactory}
   */
  @Bean
  public ConcurrentKafkaListenerContainerFactory<String, ImageStatusMessage>
          kafkaListenerContainerFactory() {
    ConcurrentKafkaListenerContainerFactory<String, ImageStatusMessage> factory =
        new ConcurrentKafkaListenerContainerFactory<>();
    factory.getContainerProperties().setAckMode(ContainerProperties.AckMode.MANUAL);
    factory.setAutoStartup(true);
    factory.setConcurrency(2);
    factory.setConsumerFactory(consumerFactory());
    factory.setCommonErrorHandler(errorHandler());
    return factory;
  }

  private <K, V> ConsumerFactory<K, V> consumerFactory() {
    var props = properties.buildConsumerProperties(null);

    props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
    props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, JsonDeserializer.class);

    props.put(ConsumerConfig.ISOLATION_LEVEL_CONFIG, "read_committed");
    props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
    props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
    props.put(ConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG,
        RoundRobinAssignor.class.getName());
    props.put(JsonDeserializer.VALUE_DEFAULT_TYPE, ImageStatusMessage.class);
    props.put(JsonDeserializer.USE_TYPE_INFO_HEADERS, false);

    return new DefaultKafkaConsumerFactory<>(props);
  }

  private DefaultErrorHandler errorHandler() {
    var fixedBackOff = new FixedBackOff(1000, 3);
    return new DefaultErrorHandler((consumerRecord, exception)
        -> log.info("Consumer error for record: {}", consumerRecord), fixedBackOff);
  }
}

File: image-processor/src/main/kotlin/org/judexmars/imageprocessor/config/kafka/KafkaConfiguration.kt
package org.judexmars.imageprocessor.config.kafka

import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.apache.kafka.clients.consumer.RoundRobinAssignor
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.clients.producer.RoundRobinPartitioner
import org.apache.kafka.common.serialization.StringDeserializer
import org.apache.kafka.common.serialization.StringSerializer
import org.judexmars.imageprocessor.config.ProcessorProperties
import org.judexmars.imageprocessor.dto.ImageStatusMessage
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.boot.autoconfigure.kafka.KafkaProperties
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory
import org.springframework.kafka.core.ConsumerFactory
import org.springframework.kafka.core.DefaultKafkaConsumerFactory
import org.springframework.kafka.core.DefaultKafkaProducerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.core.ProducerFactory
import org.springframework.kafka.listener.ContainerProperties.AckMode
import org.springframework.kafka.listener.DefaultErrorHandler
import org.springframework.kafka.support.serializer.JsonDeserializer
import org.springframework.kafka.support.serializer.JsonSerializer
import org.springframework.util.backoff.FixedBackOff

@Configuration
@EnableConfigurationProperties(KafkaProperties::class)
class KafkaConfiguration(
    private val properties: KafkaProperties,
) {
    companion object {
        val log: Logger = LoggerFactory.getLogger(KafkaConfiguration::class.java)
    }

    @Bean
    fun producerFactory(): ProducerFactory<String, ImageStatusMessage> {
        val props = properties.buildProducerProperties(null).toMutableMap()

        props[ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG] = StringSerializer::class.java
        props[ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG] = JsonSerializer::class.java
        props[ProducerConfig.PARTITIONER_CLASS_CONFIG] = RoundRobinPartitioner::class.java
        props[JsonSerializer.ADD_TYPE_INFO_HEADERS] = false
        props[ProducerConfig.LINGER_MS_CONFIG] = 0
        props[ProducerConfig.ACKS_CONFIG] = "all"
        props[ProducerConfig.RETRIES_CONFIG] = 3

        return DefaultKafkaProducerFactory(props)
    }

    @Bean
    fun kafkaTemplate(): KafkaTemplate<String, ImageStatusMessage> {
        return KafkaTemplate(producerFactory()).apply { setObservationEnabled(true) }
    }

    @Bean
    fun consumerFactory(processorProperties: ProcessorProperties): ConsumerFactory<String, ImageStatusMessage> {
        val props = properties.buildConsumerProperties(null).toMutableMap()

        props[ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG] = StringDeserializer::class.java
        props[ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG] = JsonDeserializer::class.java
        props[JsonDeserializer.VALUE_DEFAULT_TYPE] = ImageStatusMessage::class.java
        props[JsonDeserializer.USE_TYPE_INFO_HEADERS] = false
        props[ConsumerConfig.AUTO_OFFSET_RESET_CONFIG] = "earliest"
        props[ConsumerConfig.GROUP_ID_CONFIG] = processorProperties.group
        props[ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG] = false
        props[ConsumerConfig.ISOLATION_LEVEL_CONFIG] = "read_committed"
        props[ConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG] = listOf(RoundRobinAssignor::class.java)
        return DefaultKafkaConsumerFactory(props)
    }

    @Bean
    fun kafkaListenerContainerFactory(
        processorProperties: ProcessorProperties,
    ): ConcurrentKafkaListenerContainerFactory<String, ImageStatusMessage> {
        return ConcurrentKafkaListenerContainerFactory<String, ImageStatusMessage>().apply {
            with(containerProperties) {
                ackMode = AckMode.MANUAL
                setAutoStartup(true)
                setConcurrency(processorProperties.concurrency)
                setCommonErrorHandler(errorHandler())
                isObservationEnabled = true
                consumerFactory = consumerFactory(processorProperties)
            }
        }
    }

    private fun errorHandler(): DefaultErrorHandler {
        val fixedBackOff = FixedBackOff(1000, 3)
        return DefaultErrorHandler({ consumerRecord: ConsumerRecord<*, *>?, exception: Exception? ->
            log.info(
                "Consumer error for record: {}\n{}",
                consumerRecord,
                exception.toString(),
            )
        }, fixedBackOff)
    }
}
